pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,498 @@
+		// Write out the text version for debugging purposes only
+		Data: map[string]string{"targets.txt": r.targetsConfig.String()},
+	}
+
+	r.Logger.Debug("Current targets config", zap.Any("targetsConfig", r.targetsConfig.String()))
+
+	existing, err := r.configMapLister.ConfigMaps(desired.Namespace).Get(desired.Name)
+	if errors.IsNotFound(err) {
+		r.Logger.Debug("Creating targets ConfigMap", zap.String("namespace", desired.Namespace), zap.String("name", desired.Name))
+		_, err = r.KubeClientSet.CoreV1().ConfigMaps(desired.Namespace).Create(desired)
+		if err != nil {
+			return fmt.Errorf("error creating targets ConfigMap: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("error getting targets ConfigMap: %w", err)
+	}
+
+	r.Logger.Debug("Compare targets ConfigMap", zap.Any("existing", base64.StdEncoding.EncodeToString(existing.BinaryData[targetsCMKey])), zap.String("desired", base64.StdEncoding.EncodeToString(desired.BinaryData[targetsCMKey])))
+	if !equality.Semantic.DeepEqual(desired.BinaryData, existing.BinaryData) {
+		r.Logger.Debug("Updating targets ConfigMap")
+		_, err = r.KubeClientSet.CoreV1().ConfigMaps(desired.Namespace).Update(desired)
+		if err != nil {
+			return fmt.Errorf("error updating targets ConfigMap: %w", err)
+		}
+	}
+	return nil
+}
+
+// LoadTargetsConfig retrieves the targets ConfigMap and
+// populates the targets config struct.
+func (r *Reconciler) LoadTargetsConfig(ctx context.Context) error {
+	r.Logger.Debug("Loading targets config from configmap")
+	//TODO should we use the apiserver here?
+	// kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(targetsCMName. metav1.GetOptions{})
+	//TODO retry with wait.ExponentialBackoff
+	existing, err := r.configMapLister.ConfigMaps(system.Namespace()).Get(targetsCMName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.targetsConfig = memory.NewEmptyTargets()
+			return nil
+		}
+		return fmt.Errorf("error getting targets ConfigMap: %w", err)
+	}
+
+	targets := &config.TargetsConfig{}
+	data := existing.BinaryData[targetsCMKey]
+	if err := proto.Unmarshal(data, targets); err != nil {
+		return err
+	}
+
+	r.targetsConfig = memory.NewTargets(targets)
+	r.Logger.Debug("Loaded targets config from ConfigMap", zap.String("resourceVersion", existing.ResourceVersion))
+	return nil
+}
+
+func (r *Reconciler) TargetsConfigUpdater(ctx context.Context) {
+	r.Logger.Debug("Starting TargetsConfigUpdater")
+	// check every 10 seconds even if no reconciles have occurred
+	ticker := time.NewTicker(targetsCMResyncPeriod)
+
+	//TODO configmap cleanup: if any brokers are in deleted state with no triggers
+	// (or all triggers are in deleted state), remove that entry
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.Logger.Debug("Stopping TargetsConfigUpdater")
+			return
+		case <-r.targetsNeedsUpdate:
+			if err := r.updateTargetsConfig(ctx); err != nil {
+				r.Logger.Error("Error in TargetsConfigUpdater: %w", err)
+			}
+		case <-ticker.C:
```

Comment:
maybe this should start a new timer on every loop to avoid spurious updates when targets have been recently updated?

Reply:
I think this will be fixed by moving the configmap update to a different controller for #852.